server/remote_cache/cachetools/cachetools.go

```diff
@@ -272,6 +339,25 @@ func UploadFromReader(ctx context.Context, bsClient bspb.ByteStreamClient, r *di
 	return r.GetDigest(), nil
 }
 
+func UploadFromReader(ctx context.Context, bsClient bspb.ByteStreamClient, r *digest.ResourceName, in io.Reader) (*repb.Digest, error) {
+	// We can only retry if we can rewind the reader back to the beginning.
+	seeker, retryable := in.(io.Seeker)
+	if retryable {
+		return retry.Do(ctx, retryOptions("ByteStream.Write"), func(ctx context.Context) (*repb.Digest, error) {
+			if _, err := seeker.Seek(0, io.SeekStart); err != nil {
+				return nil, retry.NonRetryableError(err)
+			}
+			// We set a timeout for the entire write since we can't monitor
```

Comment:
hm can we do the same thing in UploadFromReader that we do in GetBlob, with the goroutine + timer? (I get that `stream.Send()` may buffer but isn't there some max number of outstanding unack'd messages?)

Reply:
Good idea, done. (Individual messages are not acked, but there's only so much data a sender is allowed to send before the peer has to indicate that it is willing to receive X more bytes.)